### DISCUSSION TITLE
Scheduling improvement

### DISCUSSION THREAD

**Discussion by jsa91** [micropython/micropython#13394] [ESP32] [answered]
Hi,

I'd like some feedback on my `while` loop. I'm using a driver to display clock on a ILI9341 display which is working fine. At certain times I'd like to be able to call my functions like this:

```python
while True:
    t = time.localtime()
    hrs.value(hstart * uv(-t[3]*pi/6 - t[4]*pi/360), YELLOW)
    mins.value(mstart * uv(-t[4] * pi/30), YELLOW)
    secs.value(sstart * uv(-t[5] * pi/30), RED)
    dial.text('{} {} {}'.format(days[t[6]], t[2], months[t[1] - 1]))
    refresh(ssd)
   
    hh = t[3]
    mm = t[4]
    ss = t[5]
   
    if hh == 0 and mm == 0 and ss == 0:
        my_first_function()
       
    if hh == 14 and mm == 0 and ss == 0:
        my_second_function()
    
    if hh != current_hour:
        my_third_function()
    
    current_hour = hh

    time.sleep(1)
```

I have a feeling that this is not so effective. How can I improve my code?

**Comment by robert-hh** [comment]
There is just little room for improvement. The `if` clauses may be partially exclusive, at least the first two. So you could use `elif` for the second and optionally the third `if`.
I see a different problem in this code. The execution of the loop content takes a while. So it may happen that the situation `ss == 0` is missed. That will happen, if `time.localtime()` is called at the end of a 59th second, and the time for the loop and `time.sleep(1)` adds up to more than 1 second. Then the next ss value will be >0. You shoudl see that already in the fact, that the second increases sometimes by 2 or more.
So instead of `ss == 0` you need a kind a state symbol, which is e.g. set `True` when  e.g. `ss < 3` and set `False` when e.g. `ss > 30`, and replace `ss == 0` with a test of this symbol is True.

**Reply by robert-hh** [reply, marked as answer]
or it could be something like:
```
current_hour = time.localtime()[3]
while True:
    t = time.localtime()
    hrs.value(hstart * uv(-t[3]*pi/6 - t[4]*pi/360), YELLOW)
    mins.value(mstart * uv(-t[4] * pi/30), YELLOW)
    secs.value(sstart * uv(-t[5] * pi/30), RED)
    dial.text('{} {} {}'.format(days[t[6]], t[2], months[t[1] - 1]))
    refresh(ssd)
   
    hh = t[3]

    if hh != current_hour:
        if hh == 0:
            my_first_function()
        elif hh == 14:
            my_second_function()
        else:
            my_third_function()

    current_hour = hh

    time.sleep(1)
```
The check for mm and ss is not needed, since `current_hour != hh` is only valid at the start of an hour.

**Reply by jsa91** [reply]
Thank you for your answer @robert-hh! That code seem to work exactly like how I want it to.

**Comment by bulletmark** [comment]
This is a primitive way to schedule stuff though. Better to use something like [mcron](https://github.com/fizista/micropython-mcron) or [micropython async crontab](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/SCHEDULE.md).

**Reply by jsa91** [reply]
Okey, I'll check that out. Thank you!
